[vulkan] support add for dim < 4

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #41435 WIP - vulkan view
* #41434 [vulkan] cat op (concatenate)
* #41380 [vulkan] inplace add_, relu_
* #41379 [vulkan] max_pool2d
* #41224 [vulkan][asan] Fix Invalid Memory ops
* #41223 [vulkan] reshape op
* **#41222 [vulkan] support add for dim < 4**
* #41221 [vulkan] mm op through addmm

Differential Revision: [D22754937](https://our.internmc.facebook.com/intern/diff/D22754937)

IvanKobzarev committed:
[vulkan] support add for dim < 4

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] support add for dim < 4"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] support add for dim < 4"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] support add for dim < 4"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] support add for dim < 4"


Differential Revision: [D22754937](https://our.internmc.facebook.com/intern/diff/D22754937)

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanOps.cpp:
```diff
@@ -125,19 +125,34 @@ void add(
+  std::array<int64_t, 4> i0s4 = {1, 1, 1, 1};
+  std::copy(i0s.begin(), i0s.end(), i0s4.end() - i0dim);
+  std::array<int64_t, 4> i1s4 = {1, 1, 1, 1};
+  std::copy(i1s.begin(), i1s.end(), i1s4.end() - i1dim);
```
Interesting implementation. :)

AshkanAliabadi replied:
Also you could use cbegin() and cend() to get a const iterator.  Only applicable to the source parameters.

IvanKobzarev committed:
Update on "[vulkan] support add for dim < 4"


Differential Revision: [D22754937](https://our.internmc.facebook.com/intern/diff/D22754937)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] support add for dim < 4"


Differential Revision: [D22754937](https://our.internmc.facebook.com/intern/diff/D22754937)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] support add for dim < 4"


Differential Revision: [D22754937](https://our.internmc.facebook.com/intern/diff/D22754937)

[ghstack-poisoned]
